migrated spawning processes (#3392)
jenkins-infra/jenkins.io: content/doc/book/managing/spawning-processes.adoc

---
layout: section
---
ifdef::backend-html5[]
:notitle:
:description:
:author:
:email: jenkinsci-docs@googlegroups.com
:sectanchors:
:toc:
ifdef::env-github[:imagesdir: ../resources]
ifndef::env-github[:imagesdir: ../../resources]
:hide-uri-scheme:
endif::[]

= Spawning Processes From Build

> Sometimes you'd like to spawn a process from a build that lives longer than the build itself. For example, maybe a part of the build is to launch a new application server with the result of the build. When you do this, you often experience a problem where the build doesn't terminate; you'll see that shell script/ant/maven terminates as expected, but Jenkins just insists on waiting, as if it didn't notice that the build is over.

Reviewer: ```suggestion
It is possible to spawn a process from a build and have that process live longer than the build itself.
For example, perhaps the build launches a new application server with the result of the build.
In older releases, the build often did not terminate.
Instead, the specific step (such as the shell script, Ant, or Maven) terminates
but the build itself does not terminate.
```

> Starting Jenkins 1.136, Jenkins detects this situation and instead of causing infinite block, it will just print out a warning and let you get going. But you should still understand what's causing this.

Reviewer: ```suggestion
Jenkins 1.136 and later detects this situation and, instead of blocking indefinitely,  prints out a warning and terminates the build.
```

Reviewer: Do we need to preserve the Jenkins version where this was first implemented?  It now sees so old that it is not relevant.

 
== Why?

> The reason this problem happens is because of file descriptor leak and how they are inherited from one process to another. Jenkins and the child process are connected by three pipes (stdin/stdout/stderr.) This allows Jenkins to capture the output from the child process. Since the child process may write a lot of data to the pipe and quit immediately after that, Jenkins needs to make sure that it drained the pipes before it considers the build to be over. Jenkins does this by waiting for EOF.

Reviewer: ```suggestion
This happens because of file descriptor leak and how they are inherited from one process to another.
Jenkins and the child process are connected by three pipes ('stdin', 'stdout`, and `stderr`.)
This allows Jenkins to capture the output from the child process.
Since the child process may write a lot of data to the pipe and quit immediately after that, Jenkins waits for end-of-file (EOF) to be sure that it has drained the pipes before it terminates the build. 
```

> When a process terminates for whatever reasons, the operating system closes all the file descriptors it owned. So even if the process didn't close stdout/stderr, Jenkins will nevertheless get EOF.

Reviewer: ```suggestion
Whenever a process terminates, the operating system closes all the file descriptors it owned. So, even if the process did not close `stdout` and `stder`r, Jenkins gets EOF.
```

> The complication happens when those file descriptors are inherited to other processes. Let's say the child process forks another process to the background. The background process (AKA daemon) inherits all the file descriptors of the parent, including the writing side of the stdout/stderr pipes that connect the child process and Jenkins. If the daemon forgets to close them, Jenkins won't get EOF for pipes even when the child process exits, because daemon still have those descriptors open. That's how this problem happens.

Reviewer: ```suggestion
The complication happens when those file descriptors are inherited by other processes.
Let's say the child process forks another process to the background.
The background process (which is actually a daemon) inherits all the file descriptors of the parent, including the writing side of the s`tdout` ad `stderr` pipes that connect the child process and Jenkins.
If the daemon forgets to close them, Jenkins does not get EOF for pipes even when the child process exits, because the daemon still has those descriptors open.
This is how this problem happens.
```

> A good daemon program closes all file descriptors to avoid problems like this, but often there are bad ones that don't follow the rule.

Reviewer: ```suggestion
A daemon should close all file descriptors to avoid such issues but some daemons do not follow the rule.
```

== Workarounds

> On Unix, you can use a wrapper like http://www.clapper.org/software/daemonize/[this] to make the daemon behave. You can call your command like this:

Reviewer: ```suggestion
On Unix, you can use a wrapper like http://www.clapper.org/software/daemonize/[this] to make the daemon behave.
You can call your command like this:
```

----
daemonize -E BUILD_ID=dontKillMe /path/to/your/command
----

> In case of Jenkins Pipeline use `+JENKINS_NODE_COOKIE+` instead of `+BUILD_ID+`

Reviewer: ```suggestion
In a Jenkins Pipeline, use `+JENKINS_NODE_COOKIE+` instead of `+BUILD_ID+`
```

> Note that this will set the BUILD_ID environment variable for the process being spawned to something other than the current BUILD_ID. Or you can start jenkins with -Dhudson.util.ProcessTree.disable=true - see https://wiki.jenkins.io/display/JENKINS/ProcessTreeKiller[ProcessTreeKiller] for details.

Reviewer: ```suggestion
Note that this sets the BUILD_ID environment variable for the process being spawned to something other than the current BUILD_ID.
Alternately, you can start jenkins with

----
-Dhudson.util.ProcessTree.disable=true
----
See https://wiki.jenkins.io/display/JENKINS/ProcessTreeKiller[ProcessTreeKiller] for details.
```

> On Windows, http://www.microsoft.com/resources/documentation/windows/xp/all/proddocs/en-us/ntcmds.mspx?mfr=true[the 'at' command] can be used to launch a process in the background. See the example below:

Reviewer: ```suggestion
On Windows, use the http://www.microsoft.com/resources/documentation/windows/xp/all/proddocs/en-us/ntcmds.mspx?mfr=true['at' command] to launch a process in the background.
For example:
```

....
<scriptdef name="get-next-minute" language="beanshell">
  <attribute name="property" />

  date = new java.text.SimpleDateFormat("HH:mm")
    .format(new Date(System.currentTimeMillis() + 60000));
  project.setProperty(attributes.get("property"), date);
</scriptdef>

<get-next-minute property="next-minute" />
<exec executable="at">
  <arg value="${next-minute}" />
  <arg value="/interactive" />
  <arg value="${jboss.home}\bin\run.bat" />
</exec>
....

> Another similar workaround on Windows is to use wrapper script and launch your program through it.

Reviewer: ```suggestion
Another similar workaround on Windows is to use wrapper script and launch your program through it:
```

// On Windows NT/2000/XP, specify a title for the window.  If the environment
// variable ANTRUN_TITLE is specified, that will be used instead of a default.
if (isWindowsNT) {
  if (windowTitle == "")
     windowTitle = "Ant - " + WScript.Arguments(i);
  exeStr += "title " + windowTitle + " &&";
}

// Loop through arguments quoting ones with spaces
for (var i = 0; i < WScript.Arguments.count(); i++) {
  arg = WScript.Arguments(i);
  if (arg.indexOf(' ') > 0)
    exeStr += " \"" + arg + "\"";
  else
    exeStr += " " + arg;
}

// If the environment variable ANTRUN_OUTPUT was specified, redirect
// output to that file.
if (outputFile != "") {
  windowStyle = 7;  // new window is minimized
  exeStr += " > \"" + outputFile + "\"";
  if (isWindowsNT)
    exeStr += " 2>&1";
}

// WScript.Echo(exeStr);
// WshShell.Run(exeStr);
WshShell.Run(exeStr, windowStyle, false);
....

....
<exec executable="cscript.exe">
   <env key="ANTRUN_TITLE" value="Title for Window" />  <!-- optional -->
   <env key="ANTRUN_OUTPUT" value="output.log" />  <!-- optional -->
   <arg value="//NoLogo" />
   <arg value="antRunAsync.js" />  <!-- this script -->
   <arg value="real executable" />
</exec>
....

> Another workaround for Windows XP and later is to shedule permanent task and force running it from the ant script. Once run the command:

Reviewer: ```suggestion
Another workaround for Windows XP and later is to schedule a permanent task and force running it from the Ant script.
For example, run the command:
```

....
C:\>SCHTASKS /Create /RU SYSTEM /SC ONSTART /TN Tomcat /TR 
"C:\Program Files\Apache Software Foundation\Tomcat 6.0\bin\startup.bat"
....

> Note, that ONSTART can be replaced with ONCE if you do not want to keep Tomcat running. Add the following code to your ant script:

Reviewer: ```suggestion
Note, that ONSTART can be replaced with ONCE if you do not want to keep Tomcat running.
Add the following code to your Ant script:
```

....
<exec executable="SCHTASKS">
    <arg value="/Run"/>
    <arg value="/TN"/>
    <arg value="Tomcat"/>
</exec>
....

> Another possibility that we can consider is to do something in Jenkins.

Reviewer: ```suggestion

```

When a process terminates for whatever reasons, the operating system closes all the file descriptors it owned. So even if the process didn't close stdout/stderr, Jenkins will nevertheless get EOF.

The complication happens when those file descriptors are inherited to other processes. Let's say the child process forks another process to the background. The background process (AKA daemon) inherits all the file descriptors of the parent, including the writing side of the stdout/stderr pipes that connect the child process and Jenkins. If the daemon forgets to close them, Jenkins won't get EOF for pipes even when the child process exits, because daemon still have those descriptors open. That's how this problem happens.

A good daemon program closes all file descriptors to avoid problems like this, but often there are bad ones that don't follow the rule.

== Workarounds

On Unix, you can use a wrapper like http://www.clapper.org/software/daemonize/[this] to make the daemon behave. You can call your command like this:

----
daemonize -E BUILD_ID=dontKillMe /path/to/your/command
----

In case of Jenkins Pipeline use `+JENKINS_NODE_COOKIE+` instead of `+BUILD_ID+`

> Note that this will set the BUILD_ID environment variable for the process being spawned to something other than the current BUILD_ID. Or you can start jenkins with -Dhudson.util.ProcessTree.disable=true - see https://wiki.jenkins.io/display/JENKINS/ProcessTreeKiller[ProcessTreeKiller] for details.

Reviewer 2: If https://github.com/jenkins-infra/jenkins.io/pull/3364/files#diff-f5e459407b2ff2162c1f1176c3bd543a gets merged then this link could be changed as

```suggestion
Note that this will set the BUILD_ID environment variable for the process being spawned to something other than the current BUILD_ID. Or you can start jenkins with -Dhudson.util.ProcessTree.disable=true - see link:/doc/book/using/long-running-agent-process[long running agent process] for details.
```

Author: Should I directly update this. OR wait for the other pull request to be merged before updating this.

On Windows, http://www.microsoft.com/resources/documentation/windows/xp/all/proddocs/en-us/ntcmds.mspx?mfr=true[the 'at' command] can be used to launch a process in the background. See the example below:

....
<scriptdef name="get-next-minute" language="beanshell">
  <attribute name="property" />

  date = new java.text.SimpleDateFormat("HH:mm")
    .format(new Date(System.currentTimeMillis() + 60000));
  project.setProperty(attributes.get("property"), date);
</scriptdef>

<get-next-minute property="next-minute" />
<exec executable="at">
  <arg value="${next-minute}" />
  <arg value="/interactive" />
  <arg value="${jboss.home}\bin\run.bat" />
</exec>
....

> Another similar workaround on Windows is to use https://wiki.jenkins.io/download/attachments/1835010/antRunAsync.js?version=1&modificationDate=1184046328000&api=v2[a wrapper script] and launch your program through it.

Reviewer 2: Could the reference to `https://wiki.jenkins.io/download/attachments/1835010/antRunAsync.js?version=1&modificationDate=1184046328000&api=v2` be changed?  Maybe a code snippet here should be enough?

Author: I have replaced the link with the code itself.
